### PR TITLE
Offline KD skip write teacher checkpoint when saving logits

### PR DESCRIPTION
**Pull request by jenchen13**
- [X] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
When running offline KD's first step to save logits, currently it saves the teacher checkpoint every `--save-interval`, which adds unnecessary checkpoints on disk. Skip saving the teacher checkpoint when saving logits. 

Also fixes Pipeline parallelism for offline KD, currently all ranks in PP are inspected for logits, causing a deadlock.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/datasets/data_samplers.py [hunk 1]
```diff
@@ -21,6 +21,12 @@ def build_pretraining_data_loader(dataset, consumed_samples):
 
     if dataset is None:
         return None
+    # Empty split (e.g. valid/test when --eval-iters 0): return null loader
```
Comment (AAnoosheh) [inline megatron/training/datasets/data_samplers.py:24]: Why this?

Comment (jenchen13) [reply megatron/training/datasets/data_samplers.py:24]: I want to skip validation when dumping logits but that leads to building a 0 length dataloader which will crash training. instead returning a null dataloader doesn't crash

**Comment by jenchen13** [comment]
/ok to test [fb8244d](https://github.com/NVIDIA/Megatron-LM/pull/5851/commits/fb8244d4e07796ee6c879efeb5f659d0b5189bbf)

File: megatron/training/datasets/data_samplers.py (hunk 1 above)
Comment (AAnoosheh) [reply megatron/training/datasets/data_samplers.py:24]: I just set validation interval to something super large and that works fine

File: megatron/training/distillation/cached_logits_loss.py [hunk 2]
```diff
@@ -105,6 +105,10 @@ def __init__(self):
 
     def attach_hooks(self, model: LanguageModule) -> None:
         """Attach forward hooks to the model's output layer."""
+        # Only the last pipeline stage has output_layer (post_process=True); guard so PP>1
```
Comment (AAnoosheh) [inline megatron/training/distillation/cached_logits_loss.py:108]: Will be fixed in https://github.com/NVIDIA/Megatron-LM/pull/5926

File: megatron/training/distillation/logits_saver.py [hunk 3]
```diff
@@ -252,6 +252,9 @@ def attach_hooks(self, model: LanguageModule) -> None:
         Args:
             model: Model to instrument.
         """
+        # Only the last pipeline stage has output_layer / computes logits
```
Comment (AAnoosheh) [inline megatron/training/distillation/logits_saver.py:255]: Will be fixed in https://github.com/NVIDIA/Megatron-LM/pull/5926

File: megatron/training/training.py [hunk 4]
```diff
@@ -2360,7 +2360,7 @@ def train_step(forward_step_func, data_iterator, model, optimizer, opt_param_sch
             seq_length=args.seq_length,
             micro_batch_size=args.micro_batch_size,
             decoder_seq_length=args.decoder_seq_length,
-            forward_only=False,
+            forward_only=args.freeze_all_layers,
```
Comment (AAnoosheh) [inline megatron/training/training.py:2363]: Will be fixed in https://github.com/NVIDIA/Megatron-LM/pull/5926

File: pretrain_gpt.py [hunk 5]
```diff
@@ -289,7 +289,9 @@ def forward_step(data_iterator, model: GPTModel, return_schedule_plan: bool = Fa
 
     # Get the batch.
     timers('batch-generator', log_level=2).start()
-    with stimer(bdata=True):
+    # Frozen teacher (--freeze-all-layers) runs forward-only for logit saving: disable grad
```
Comment (AAnoosheh) [inline pretrain_gpt.py]: Will be fixed in https://github.com/NVIDIA/Megatron-LM/pull/5926

File: pretrain_hybrid.py [hunk 6]
```diff
@@ -328,7 +329,9 @@ def forward_step(data_iterator, model: HybridModel):
 
     timers('batch-generator').stop()
 
-    with stimer:
+    # Frozen teacher (--freeze-all-layers) runs forward-only for logit saving: disable grad
```
Comment (AAnoosheh) [inline pretrain_hybrid.py]: Will be fixed in https://github.com/NVIDIA/Megatron-LM/pull/5926

File: pretrain_gpt.py (hunk 5 above)
Comment (jenchen13) [reply pretrain_gpt.py]: I can undo the changes to pretrain_*.py but I will keep my other changes

File: megatron/training/distillation/cached_logits_loss.py (hunk 2 above)
Comment (jenchen13) [reply megatron/training/distillation/cached_logits_loss.py:108]: i'm keeping this

File: megatron/training/distillation/logits_saver.py (hunk 3 above)
Comment (jenchen13) [reply megatron/training/distillation/logits_saver.py:255]: I'm keeping this here, please rebase

**Review by cuichenx — APPROVED** [review]

File: megatron/training/distillation/cached_logits_loss.py (hunk 2 above)
Comment (AAnoosheh) [reply megatron/training/distillation/cached_logits_loss.py:108]: The reason my PR guards it in training.py is so that `get_student_logits_capture()` and `get_logits_saver()` return None. In this case, non-final PP ranks will still create an instance of it and then downstream results in incorrect behavior. So we want `attach_hooks()` to cause an error if called on non-final PP ranks.

You can also rebase on top of that small PR and drop some of the current changes here. (And I close my PR)

**Comment by AAnoosheh** [comment]
Changes migrated to https://github.com/NVIDIA/Megatron-LM/pull/5926
